web/modules/custom/brms_sor/src/Service/FmsSorApi.php

```diff
@@ -0,0 +1,252 @@
+   * @param \Drupal\key\KeyRepositoryInterface $key_repository
+   *   The key repository.
+   * @param \Drupal\Component\Serialization\Json $json_service
+   *   The JSON serializer service.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The time service.
+   */
+  public function __construct(
+    protected ClientInterface $http_client,
+    protected LoggerChannelFactoryInterface $logger_factory,
+    protected CacheBackendInterface $cache_backend,
+    protected KeyRepositoryInterface $key_repository,
+    protected Json $json_service,
+    protected TimeInterface $time,
+  ) {
+    $this->httpClient = $http_client;
+    $this->logger = $logger_factory;
+    $this->cache = $cache_backend;
+    $this->json = $json_service;
+    $settings = Settings::get('brms_sor');
+    $this->authUrl = $settings['fms_api_auth_url'];
+    $this->baseUrl = $settings['fms_api_base_url'];
+    $this->clientId = $settings['fms_api_client_id'];
+    $this->clientSecret = $key_repository->getKey('fms_api')->getKeyValue();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAccessToken(bool $use_cached_token = TRUE): string {
+    if ($use_cached_token) {
+      $cache_access_token = $this->cache->get('brms_sor_fms_access_token');
+      if ($cache_access_token) {
+        return $cache_access_token->data;
+      }
+    }
+
+    try {
+      $response = $this->httpClient->request('POST', $this->authUrl, [
+        'headers' => [
+          'Content-Type' => 'application/x-www-form-urlencoded',
+        ],
+        'form_params' => [
+          'grant_type' => 'client_credentials',
+          'client_id' => $this->clientId,
+          'client_secret' => $this->clientSecret,
+        ],
+      ]);
+
+      $data = $this->json->decode($response->getBody()->getContents());
+      $this->cache->set('brms_sor_fms_access_token', $data['access_token'], $this->time->getRequestTime() + $data['expires_in']);
+
+      return $data['access_token'];
+    }
+    catch (\Exception $e) {
+      $this->logger->get('brms_sor')->error('Failed to get access token from the FMS API: @message', ['@message' => $e->getMessage()]);
+      throw $e;
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function get(string $call, array $data = [], bool $use_cached_token = TRUE): array {
+    try {
+      $token = $this->getAccessToken($use_cached_token);
+      $headers = [
+        'Authorization' => 'Bearer ' . $token,
+      ];
+
+      $args = ['headers' => $headers, 'query' => $data];
+      $uri = $this->baseUrl . $call;
+      $this->response = $this->httpClient->get($uri, $args);
+      return $this->json->decode($this->response->getBody()->getContents());
+    }
+    catch (RequestException $e) {
+      $this->response = $e->getResponse();
+      $this->logger->get('brms_sor')->error('Failed to call GET @uri on FMS API: @message',
+        ['@uri' => $uri, '@message' => $e->getMessage()]);
+      // If the token expired, try again with a new token.
```

Comment:
We should not do this. We know if it has expired beforehand.

Reply:
fixed by checking the token expiry data above